### PR TITLE
[Web Components] Add Shadow DOM Support to CodeceptJS

### PR DESCRIPTION
**Pull request by gkushang**
## Motivation/Description of the PR
- Shadow-DOM locator integration with webdriverio 5.x. [Link](https://codecept.io/wiki/Roadmap.html#backlog) to Requested Feature on Roadmap
- Resolves https://github.com/Codeception/CodeceptJS/issues/2275 https://github.com/Codeception/CodeceptJS/issues/1003.

Applicable helpers:

- [x] WebDriver
- [ ] Puppeteer
- [ ] Nightmare
- [ ] REST
- [ ] FileHelper
- [ ] Appium
- [ ] Protractor
- [ ] TestCafe
- [ ] Playwright

Applicable plugins:

- [ ] allure
- [ ] autoDelay
- [ ] autoLogin
- [ ] customLocator
- [ ] pauseOnFail
- [ ] puppeteerCoverage
- [ ] retryFailedStep
- [ ] screenshotOnFail
- [ ] selenoid
- [ ] stepByStepReport
- [x] wdio

## Type of change

- [ ] :fire: Breaking changes
- [x] :rocket: New functionality
- [ ] :bug: Bug fix
- [x] :clipboard: Documentation changes/updates
- [ ] :hotsprings: Hot fix
- [ ] :hammer: Markdown files fix - not related to source code
- [ ] :nail_care: Polish code

## Checklist:

- [x] Tests have been added
- [x] Documentation has been added (Run `npm run docs`)
- [x] Lint checking (Run `npm run lint`)
- [x] Local tests are passed (Run `npm test`)

### CODE REVIEW

File: lib/helper/WebDriver.js [hunk 1]
```diff
@@ -640,6 +640,32 @@ class WebDriver extends Helper {
   async _locate(locator, smartWait = false) {
     if (require('../store').debugMode) smartWait = false;
 
+    // special locator type for Shadow DOM
+    if (locator.type === 'shadowDom' || locator.shadowDom) {
```
Comment (DavertMik) [inline lib/helper/WebDriver.js]: Good, please refactor it to its own custom method `_locateShadowDom()`

Comment (gkushang) [reply lib/helper/WebDriver.js]: done

**Comment by gkushang** [comment]
@DavertMik I've also added the integration test that tests for the Shadow Dom locators and actions. The integration tests are here: test/acceptance/gherkin/shadow_dom.feature.

**Comment by gkushang** [comment]
@DavertMik 

Can you take a look when you get a chance?

**Comment by DavertMik** [comment]
Thanks, a very good PR.
I have some doubts about syntax.

Can we avoid using `elements` here?

```
{ shadowDom: { elements: ['my-app', 'recipe-hello-binding', 'ui-input', 'input.input'] } }
```

Can we just use:

```
{ shadowDom: ['my-app', 'recipe-hello-binding', 'ui-input', 'input.input'] }
```

Also, what do you think of adding shadow dom support to `within` blocks?
Maybe it would be also nice to have it working like this:

```js
within({ shadowDom: 'myapp' }, () => {

});
```

What do you think?

**Comment by Georgegriff** [comment]
Hello was asked to provide some input on this in terms of playwright, (eventually) I want to look at shadow Dom support for playwright, I was considering doing this as a third party plugin to CodeceptJS https://github.com/Georgegriff/query-selector-shadow-dom/blob/master/plugins/codeceptjs/README.md

As for syntax, my approach is different, I'm making use of that library that I maintain, it makes it so you don't need to do any custom syntax or arrays of elements to pierce you just write a normal CSS selector and the library handles the rest of piercing shadow Dom roots.


The reason I've opted to do this as a plugin rather than contribute is that not everyone likes this approach. Example of what it looks like to use my lib in playwright  
```
I.amOnPage("chrome://downloads");
  I.see("Files you download appear here", {shadow: "#no-downloads span"});
  I.waitForVisible({shadow: "#no-downloads"}, 5);
  I.click({shadow: `[title="Search downloads"]`});
  I.waitForVisible({shadow: '#searchInput'}, 5);
  I.fillField({shadow: '#searchInput'}, "A download")
  I.waitForValue({shadow: '#searchInput'}, "A download", 5)
  I.waitForText("No search results found", 3, {shadow: "#no-downloads span"});
  I.clearField({shadow: '#searchInput'})
  I.waitForValue({shadow: '#searchIn
```

**Comment by gkushang** [comment]
> Thanks, a very good PR.
> I have some questions about API. Can we discuss them?
> 
> 1. Can we avoid using `elements` here?
> 
> ```
> { shadowDom: { elements: ['my-app', 'recipe-hello-binding', 'ui-input', 'input.input'] } }
> ```
> 
> Can't we just use:
> 
> ```
> { shadowDom: ['my-app', 'recipe-hello-binding', 'ui-input', 'input.input'] }
> ```
> 
> 1. Also, what do you think of adding shadow dom support to `within` blocks?
>    Maybe it would be also nice to have it working like this:
> 
> ```js
> within({ shadowDom: 'myapp' }, () => {
> 
> });
> ```
> 
> What do you think?

Sure, I can omit the `elements`. I think it's a good idea to also have support with `within`. I can take a look at it. Would it be okay if I can provide another PR with the `within` support as I do not want this PR to grow bigger? And, also we can start using in our Salsforce's LWC app? Let me know what you think?

**Comment by gkushang** [comment]
> Hello was asked to provide some input on this in terms of playwright, so we can potentially align on syntax. (eventually) I want to look at shadow Dom support for playwright, I was considering doing this as a third party plugin to CodeceptJS https://github.com/Georgegriff/query-selector-shadow-dom/blob/master/plugins/codeceptjs/README.md
> 
> As for syntax, my approach is different, I'm making use of that library that I maintain, it makes it so you don't need to do any custom syntax or arrays of elements to pierce shadow dom you just write a normal CSS selector and the library handles the rest of piercing shadow Dom roots.
> 
> The reason I've opted to do this as a plugin rather than contribute is that not everyone likes this approach. Example of what it looks like to use my lib in playwright . All of the elements are several levels deep in shadow Dom, but when writing test you don't care
> 
> ```
> I.amOnPage("chrome://downloads");
>   I.see("Files you download appear here", {shadow: "#no-downloads span"});
>   I.waitForVisible({shadow: "#no-downloads"}, 5);
>   I.click({shadow: `[title="Search downloads"]`});
>   I.waitForVisible({shadow: '#searchInput'}, 5);
>   I.fillField({shadow: '#searchInput'}, "A download")
>   I.waitForValue({shadow: '#searchInput'}, "A download", 5)
>   I.waitForText("No search results found", 3, {shadow: "#no-downloads span"});
>   I.clearField({shadow: '#searchInput'})
>   I.waitForValue({shadow: '#searchIn
> ```
> 
> If you did want to take a look at seeing if my library could work in webdriverio it essentially just replaces querySelector/querySelectorAll with one that automatically goes through shadow roots
> 
> I've integrated the library in the past in selenium using custom locator strategies, not sure if web driver Io has similar

that's cool! I will take a look.

**Comment by gkushang** [comment]
@DavertMik - I have updated the syntax to below,

```
{ shadow: ['my-app', 'recipe-hello', 'button'] }
```

Please take a look! Thanks!

File: lib/helper/WebDriver.js [hunk 2]
```diff
@@ -640,6 +697,19 @@ class WebDriver extends Helper {
...
     if (require('../store').debugMode) smartWait = false;
 
+    // special locator type for Shadow DOM
+    if (this._isShadowLocator(locator)) {
+      if (!this.options.smartWait || !smartWait) {
+        const els = await this._locateShadow(locator);
+        return els;
+      }
+
+      await this._smartWait(locator);
```
Comment (DavertMik) [inline lib/helper/WebDriver.js]: Smart wait won't work with shadow dom
```suggestion
```

**Comment by DavertMik** [comment]
Looks fine to me! Maybe there will be more comments here
Anyway I plan to include this into a next 2.5.1 release which I hope will be available tomorrow

File: lib/helper/WebDriver.js (hunk 2 above)
Comment (gkushang) [reply lib/helper/WebDriver.js]: IWhen I try it waits and also incorporates the `retry`. Pls see below, 

```
I click {"shadow":["downloads-manager","downloads-toolbar","cr-toolbar","cr-toolbar-search-field","[title=\"Search downloads\"]"]}
    › [SmartWait (5000ms)] Locating {shadow: downloads-manager,downloads-toolbar,cr-toolbar,cr-toolbar-search-field,[title="Search downloads"]} in 5000
    [1] Retrying... Attempt #2
    › [SmartWait (5000ms)] Locating {shadow: downloads-manager,downloads-toolbar,cr-toolbar,cr-toolbar-search-field,[title="Search downloads"]} in 5000
    [1] Retrying... Attempt #3
    › [SmartWait (5000ms)] Locating {shadow: downloads-manager,downloads-toolbar,cr-toolbar,cr-toolbar-search-field,[title="Search downloads"]} in 5000
    [1] Retrying... Attempt #4
    › [SmartWait (5000ms)] Locating {shadow: downloads-manager,downloads-toolbar,cr-toolbar,cr-toolbar-search-field,[title="Search downloads"]} in 5000
    [1] Retrying... Attempt #5
    › [SmartWait (5000ms)] Locating {shadow: downloads-manager,downloads-toolbar,cr-toolbar,cr-toolbar-search-field,[title="Search downloads"]} in 5000
```

**Comment by DavertMik** [comment]
@Georgegriff looks like the approach of @gkushang is more flexible in this case. Scanning all shadow dom roots for an element is time consuming. So I think we should go with the sytnax provided in this PR as default one and maybe optionally add your simplified version for Playwright.

**Comment by Georgegriff** [comment]
Makes sense that's why I added mine as a plugin rather than a direct pr to CodeceptJS the approach can be slower you're right. The reason I've used my approach in the past is writing reusable steps for components is tricky if you have to specify ever parent all the way down to the component you want to interact with

**Comment by gkushang** [comment]
@DavertMik Thanks! I am just curious to know why Smart Wait won't work with the Shadow elements? It ultimately set's the browser timeout and looks like an implicit wait?

**Comment by DavertMik** [comment]
@gkushang 

> Thanks! I am just curious to know why Smart Wait won't work with the Shadow elements? It ultimately set's the browser timeout and looks like an implicit wait?

Yes, this is just implicit wait. For what I know, to locate shadow dom elements, WebDriver API won't be used, so implicit wait won't work.

**Comment by gkushang** [comment]
> @gkushang
> 
> > Thanks! I am just curious to know why Smart Wait won't work with the Shadow elements? It ultimately set's the browser timeout and looks like an implicit wait?
> 
> Yes, this is just implicit wait. For what I know, to locate shadow dom elements, WebDriver API won't be used, so implicit wait won't work.

Okay, thanks! Let me know when it's ready to be merged!

**Comment by Georgegriff** [comment]
Fyi looks like playwright will just work with shadow Dom our of the box next release https://github.com/microsoft/playwright/pull/1784/files they have taken the same approach as my library internally in playwright where CSS selectors recursively pierce shadow roots to find elements, no special syntax required. So from a CodeceptJS POV for playwright nothing needs to be done for shadow Dom support

**Comment by JosefJezek** [comment]
@gkushang could you add support for playwright?

**Comment by Georgegriff** [comment]
> @gkushang could you add support for playwright?

There is no need, playwright supports shadow dom out of the box it automatically pierces shadow roots, you don't need anything special

**Pull request merged**